pkg/correlation/hook.go

```diff
@@ -0,0 +1,157 @@
+		// is provided, use that context.
+		ctx = lh.FunctionContext
+	}
+
+	// If a context has been found, we will populate the correlation info
+	if ctx != nil {
+		correlationContext, ok := ctx.Value(ContextKey).(*RequestContext)
+		if !ok {
+			return fmt.Errorf("failed to get context for correlation logging hook")
+		}
+
+		entry.Data[LogFieldID] = correlationContext.ID
+		entry.Data[LogFieldOrigin] = correlationContext.Origin
+	}
+
+	// Add component when provided as a hook
+	if len(lh.Component) > 0 {
+		entry.Data[LogFieldComponent] = lh.Component
+	} else if entry.HasCaller() {
+		// Discover the component based on which package directories have
+		// been registered as hooks
+		dir := filepath.Dir(entry.Caller.File)
+		if comp, ok := components[dir]; ok {
+			entry.Data[LogFieldComponent] = comp
+		} else {
+			// If component is not registered or provided, we
+			// can add more context by looking at the caller dir
+			entry.Data[LogFieldComponent] = getLocalPackage(dir)
+		}
+
+	}
+
+	return nil
+}
+
+// NewPackageLogger creates a package-level logger for a given component
+func NewPackageLogger(component Component) *logrus.Logger {
+	clogger := logrus.New()
+	clogger.AddHook(&LogHook{
+		Component: component,
+	})
+
+	return clogger
+}
+
+// NewFunctionLogger creates a logger for usage at a per-function level
+// For example, this logger can be instantiated inside of a function with a given
+// context object. As logs are printed, they will automatically include the correlation
+// context info.
+func NewFunctionLogger(ctx context.Context) *logrus.Logger {
+	clogger := logrus.New()
+	clogger.AddHook(&LogHook{
+		FunctionContext: ctx,
+	})
+
+	return clogger
+}
+
+// RegisterGlobalHook will register the correlation logging
+// hook at a global/multi-package level. Note that this is not
+// component-aware and it is better to use NewFunctionLogger
+// or NewPackageLogger for logging.
+func RegisterGlobalHook() {
+	logrus.AddHook(&LogHook{})
+}
+
+// RegisterComponent registers the package where this function was called from
+// as a given component name.
+func RegisterComponent(component Component) {
+	_, file, _, ok := runtime.Caller(1)
+	if !ok {
+		logrus.Errorf("failed to register component")
+	}
+	registerFileAsComponent(file, component)
+}
+
+func registerFileAsComponent(file string, component Component) {
+	mu.Lock()
+	defer mu.Unlock()
+	components[filepath.Dir(file)] = component
+}
+
+// takes a directory in a go path and returns the local package.
+// i.e. /go/src/github.com/libopenstorage/openstorage/pkg/correlation
+// will return openstorage/pkg/correlation
+func getLocalPackage(dir string) string {
+	parts := strings.Split(dir, "/")
+	var githubIndex int
+	for i, d := range parts {
+		// This will work for both imported versions of this
+		// as well as in openstorage itself, since we'll always
+		// keep track of the last index for "github" in the path
+		if d == "github.com" {
```

Comment:
a path could contain something other than github.com like k8s.io
In that case we will end up logging a huge package path starting from the 2nd index.

Reply:
Having said that this won't be a problem for openstorage or porx, but could be a problem for someone using this library.